src/js/components/FilterableList.jsx

```diff
@@ -0,0 +1,147 @@
+ * @property {array} [data] data to render in the list.
+ * @property {Function} [labelFunction] converts each object to a string in the default renderer.
+ * @property {Function} [keyFunction]
+ * @property {boolean} [defaultStyles] set "false" to remove all default styling from the List.
+ * @property {Object} [defaultSelection] item to select in the list by default
+ * @property {Function} [onItemSelect] callback when an item is selected, receiving listIndex and listItem.
+ *
+ * @property {Object} listStyle React style object to be applied to child List.
+ * @property {Function} filterFunction receives text, item; return true/false to show each item
+ * @property {string} placeholder text to display in filter input when empty
+ * @property {string} emptyText text to display when no matches occur
+ */
+export class FilterableList extends React.Component {
+
+    props: Props;
+    state: State;
+
+    static defaultProps: Props = {
+        style: {},
+        listStyle: {},
+        filterFunction: defaultFilterFunction,
+        placeholder: 'Search...',
+        emptyText: 'No matches.'
+    };
+
+    constructor(props:Props) {
+        super(props);
+
+        this.state = {};
+    }
+
+    _onFilterChange = debounce(text => {
+        this.setState({
+            text,
+        });
+    }, 250);
+
+    render() {
+        // passthrough props for List
+        const { keyFunction, labelFunction, listStyle, defaultStyles, defaultSelection, onItemSelect } = this.props;
+        const { data, disabled, emptyText, filterFunction, placeholder, style } = this.props;
+
+        const { text } = this.state;
+        const filtered = data ? data.filter(item => !text || filterFunction(text, item)) : [];
+        const noMatches = text && filtered.length === 0;
+
+        const listProps = {
+            labelFunction,
+            keyFunction,
+            defaultStyles,
+            defaultSelection,
+            onItemSelect,
+        };
+
+        const outerClass = this.props.className || '';
+
+        return (
+            <div className={`FilterableList ${outerClass}`} style={style}>
+                <TextInput
+                    placeholder={placeholder}
+                    disabled={disabled}
+                    onChange={text => this._onFilterChange(text)}
+                    iconLeft="search"
+                />
+
+                { noMatches && emptyText &&
+                    <div className="FilterableList-empty-text">{emptyText}</div>
+                }
+
+                { !noMatches &&
+                <List
+                    {...listProps}
+                    disabled={disabled}
+                    className="FilterableList-List"
+                    data={filtered}
+                    style={listStyle}
+                >
+                    {this.props.children}
+                </List>
+                }
+            </div>
+        );
+    }
+}
+
+FilterableList.propTypes = {
+    // from List
```

Comment:
If this isn't a subset, could we use `...List.propTypes` so they automatically stay in sync?